src/waypoint.jsx

```diff
@@ -42,34 +43,33 @@ export default class Waypoint extends React.Component {
       return;
     }
 
-    this._handleScroll = this._handleScroll.bind(this);
-    this.scrollableAncestor = this._findScrollableAncestor();
+    // this._ref may occasionally not be set at this time. To help ensure that
+    // this works smoothly, we want to delay the initial execution until the
+    // next tick.
+    this.cancelInitialTimeout = onNextTick(() => {
+      this._handleScroll = this._handleScroll.bind(this);
+      this.scrollableAncestor = this._findScrollableAncestor();
 
-    if (process.env.NODE_ENV !== 'production' && this.props.debug) {
-      debugLog('scrollableAncestor', this.scrollableAncestor);
-    }
+      if (process.env.NODE_ENV !== 'production' && this.props.debug) {
+        debugLog('scrollableAncestor', this.scrollableAncestor);
+      }
 
-    this.scrollEventListenerHandle = addEventListener(
-      this.scrollableAncestor,
-      'scroll',
-      this._handleScroll,
-      { passive: true }
-    );
+      this.scrollEventListenerHandle = addEventListener(
+        this.scrollableAncestor,
+        'scroll',
+        this._handleScroll,
+        { passive: true }
+      );
 
-    this.resizeEventListenerHandle = addEventListener(
-      window,
-      'resize',
-      this._handleScroll,
-      { passive: true }
-    );
+      this.resizeEventListenerHandle = addEventListener(
+        window,
+        'resize',
+        this._handleScroll,
+        { passive: true }
+      );
 
-    // this._ref may occasionally not be set at this time. To help ensure that
-    // this works smoothly, we want to delay the initial execution until the
-    // next tick.
-    this.initialTimeout = setTimeout(() => {
-      this.initialTimeout = null;
       this._handleScroll(null);
```

Comment:
I was thinking that, depending on what happens in `handleScroll`, there might be some layout thrashing that happens and we might be able to reduce that by moving that work out. Or, maybe we should put the calls to `onEnter`, `onLeave`, and `onPositionChange` in `handleScroll` in an `onNextTick` callback. But I worry that might cause some confusing behavior and should perhaps be left up to the consumer to avoid thrashing if it comes up.